// @antv/g2plot/lib/plots/venn/adaptor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.adaptor = exports.axis = exports.legend = exports.LEGEND_SPACE = void 0;
var tslib_1 = require("tslib");
var util_1 = require("@antv/util");
var common_1 = require("../../adaptor/common");
var geometries_1 = require("../../adaptor/geometries");
var utils_1 = require("../../utils");
var utils_2 = require("../../utils");
var utils_3 = require("./utils");
var constant_1 = require("./constant");
require("./shape");
require("./label");
require("./interaction");
/** 图例默认预留空间 */
exports.LEGEND_SPACE = 40;
/**
 * color options 转换
 */
function transformColor(params, data) {
    var chart = params.chart, options = params.options;
    var color = options.color, setsField = options.setsField;
    if (typeof color !== 'function') {
        var colorPalette_1 = typeof color === 'string' ? [color] : color;
        if (!util_1.isArray(colorPalette_1)) {
            var _a = chart.getTheme(), colors10 = _a.colors10, colors20 = _a.colors20;
            colorPalette_1 = data.filter(function (d) { return d[setsField].length === 1; }).length <= 10 ? colors10 : colors20;
        }
        var colorMap_1 = utils_3.getColorMap(colorPalette_1, data, options);
        return function (datum) { return colorMap_1.get(datum[constant_1.ID_FIELD]) || colorPalette_1[0]; };
    }
    return color;
}
/**
 * 处理 padding
 */
function padding(params) {
    var chart = params.chart, options = params.options;
    var legend = options.legend, appendPadding = options.appendPadding, padding = options.padding;
    // 处理 legend 的位置. 默认预留 40px, 业务上可以通过 appendPadding 增加
    var tempPadding = utils_1.normalPadding(appendPadding);
    if (legend !== false) {
        tempPadding = utils_1.getAdjustAppendPadding(appendPadding, util_1.get(legend, 'position'), exports.LEGEND_SPACE);
    }
    chart.appendPadding = utils_1.resolveAllPadding([tempPadding, padding]);
    return params;
}
/**
 * 处理非法数据
 * @param params
 */
function data(params) {
    var options = params.options;
    /* 如遇到 交集 中存在 非法元素 的情况，就过滤掉
     * 如：
     * data = [
     *   { sets: ['A'], size: 3 }, // 集合
     *   { sets: ['B'], size: 4 }, // 集合
     *   { sets: ['A', 'B'], size: 2 }, // 交集
     *   { sets: ['A', 'B', 'C'], size: 2 }, // 交集 (存在非法 C，过滤该条数据)
     *   ...
     * ]
     */
    var data = options['data'];
    if (!data) {
        utils_2.log(utils_2.LEVEL.WARN, false, 'warn: %s', '数据不能为空');
        data = [];
    }
    // 合法元素的集合：['A', 'B']
    var currSets = data.filter(function (datum) { return datum.sets.length === 1; }).map(function (datum) { return datum.sets[0]; });
    // 过滤 data
    var filterSets = data.filter(function (datum) {
        var sets = datum.sets;
        // 存在非法元素，就过滤这条数据
        return utils_3.islegalSets(currSets, sets);
    });
    if (!util_1.isEqual(filterSets, data))
        utils_2.log(utils_2.LEVEL.WARN, false, 'warn: %s', '交集中不能出现不存在的集合, 请输入合法数据');
    return util_1.deepMix({}, params, {
        options: {
            data: filterSets,
        },
    });
}
/**
 * geometry 处理
 * @param params
 */
function geometry(params) {
    var chart = params.chart, options = params.options;
    var pointStyle = options.pointStyle, setsField = options.setsField, sizeField = options.sizeField;
    // 获取容器大小
    var _a = utils_1.normalPadding(chart.appendPadding), t = _a[0], r = _a[1], b = _a[2], l = _a[3];
    // 处理 legend 的位置. 默认预留 40px, 业务上可以通过 appendPadding 增加
    var customInfo = { offsetX: l, offsetY: t };
    // coordinateBBox + appendPadding = viewBBox, 不需要再计算 appendPadding 部分，因此直接使用 viewBBox
    var _b = chart.viewBBox, width = _b.width, height = _b.height;
    // 处理padding输入不合理的情况， w 和 h 不能为负数
    var vennData = utils_3.layoutVennData(options, Math.max(width - (r + l), 0), Math.max(height - (t + b), 0), 0);
    chart.data(vennData);
    var ext = geometries_1.schema(utils_1.deepAssign({}, params, {
        options: {
            xField: 'x',
            yField: 'y',
            sizeField: sizeField,
            seriesField: constant_1.ID_FIELD,
            rawFields: [setsField, sizeField],
            schema: {
                shape: 'venn',
                style: pointStyle,
                color: transformColor(params, vennData),
            },
        },
    })).ext;
    var geometry = ext.geometry;
    geometry.customInfo(customInfo);
    return params;
}
/**
 * 处理 label
 * @param params
 */
function label(params) {
    var chart = params.chart, options = params.options;
    var label = options.label;
    // 获取容器大小
    var _a = utils_1.normalPadding(chart.appendPadding), t = _a[0], r = _a[1], b = _a[2], l = _a[3];
    // 传入 label 布局函数所需的 自定义参数
    var customLabelInfo = { offsetX: l, offsetY: t };
    var geometry = utils_1.findGeometry(chart, 'schema');
    if (!label) {
        geometry.label(false);
    }
    else {
        var callback = label.callback, cfg = tslib_1.__rest(label, ["callback"]);
        geometry.label({
            fields: ['id'],
            callback: callback,
            cfg: util_1.deepMix({}, utils_1.transformLabel(cfg), {
                // 使用 G2 的 自定义label 修改位置
                type: 'venn',
                customLabelInfo: customLabelInfo,
            }),
        });
    }
    return params;
}
/**
 * legend 配置
 * @param params
 */
function legend(params) {
    var chart = params.chart, options = params.options;
    var legend = options.legend, sizeField = options.sizeField;
    chart.legend(constant_1.ID_FIELD, legend);
    // 强制不开启 连续图例
    chart.legend(sizeField, false);
    return params;
}
exports.legend = legend;
/**
 * 默认关闭坐标轴
 * @param params
 */
function axis(params) {
    var chart = params.chart;
    chart.axis(false);
    return params;
}
exports.axis = axis;
/**
 * 图适配器
 * @param chart
 * @param options
 */
function adaptor(params) {
    // flow 的方式处理所有的配置到 G2 API
    return utils_1.flow(padding, common_1.theme, data, geometry, label, common_1.scale({}), legend, axis, common_1.tooltip, common_1.interaction, common_1.animation
    // ... 其他的 adaptor flow
    )(params);
}
exports.adaptor = adaptor;
